### PR TITLE
Financial Connections: Implemented support for Multi-Select in networking (data flows)

### PR DESCRIPTION
**Pull request by kgaidis-stripe**
## Summary

This PR implements support for multi-select in Financial Connections Networking(/Link) flows.

Previously, you only were able to select a single account, now you can select multiple. 

This PR has a follow-up PR here:
- https://github.com/stripe/stripe-ios/pull/3459

### Account Required Drawer

Simulator Screenshot - iPhone 15 Pro - 2024-04-01 at 14 16 28

### LinkAccountPicker Multi Selecting (and other interaction)

https://github.com/stripe/stripe-ios/assets/105514761/37d9b43e-c148-4da1-a7d3-0e323cf99cd8

### Completing The Flow (Connecting 2 Accounts)

https://github.com/stripe/stripe-ios/assets/105514761/2327f801-a99d-496f-a73d-105960909822

### Completing The Flow (Handling Update Required ACcount Which Will Connect Only 1 Of Them)

https://github.com/stripe/stripe-ios/assets/105514761/21a5b17f-5cae-4957-a4fc-05a10299f1c3

### CODE REVIEW

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift [hunk 1]
```diff
@@ -387,11 +387,52 @@ final class AccountPickerViewController: UIViewController {
...
+                    // additional step of saving the accounts to Link
+                    if
+                        self.dataSource.manifest.isNetworkingUserFlow == true,
+                        // make sure the current user is a Link consumer
+                        self.dataSource.manifest.accountholderIsLinkConsumer == true,
+                        // make sure this is not a payment flow; in payment flows,
+                        // the AttachLinkedPaymentAccount handles saving to link
+                        !self.dataSource.manifest.shouldAttachLinkedPaymentMethod,
+                        selectedAccounts.count > 0,
+                        let consumerSessionClientSecret = self.dataSource.consumerSessionClientSecret
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift:401]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/ba55e4ac/src/linkedAccounts/inner/components/panes/v3/AccountPickerPane/lib/useSelectAccountsAndPossiblySaveToLink.ts#L86-L92)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift [hunk 2]
```diff
@@ -387,11 +387,52 @@ final class AccountPickerViewController: UIViewController {
...
+                    {
+                        self.dataSource.saveAccountsToLink(
+                            consumerSessionClientSecret: consumerSessionClientSecret
+                        )
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift:405]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/ba55e4ac/src/linkedAccounts/inner/hooks/v3/useSaveToNetworkAndLink.ts#L101-L122)

(note that this is missing `pollAccountNumbersForSelectedAccounts` which will be tackled in another PR)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift [hunk 3]
```diff
@@ -387,11 +387,52 @@ final class AccountPickerViewController: UIViewController {
...
+                                        pane: .accountPicker
+                                    )
+                            }
+
+                            self.delegate?.accountPickerViewController(
+                                self,
+                                didSelectAccounts: selectedAccounts,
+                                nextPane: .success
+                            )
+                        }
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift:424]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/ba55e4ac/src/linkedAccounts/inner/components/panes/v3/AccountPickerPane/lib/useSelectAccountsAndPossiblySaveToLink.ts#L53-L55)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift [hunk 4]
```diff
@@ -387,11 +387,52 @@ final class AccountPickerViewController: UIViewController {
...
+                        // the AttachLinkedPaymentAccount handles saving to link
+                        !self.dataSource.manifest.shouldAttachLinkedPaymentMethod,
+                        selectedAccounts.count > 0,
+                        let consumerSessionClientSecret = self.dataSource.consumerSessionClientSecret
+                    {
+                        self.dataSource.saveAccountsToLink(
+                            consumerSessionClientSecret: consumerSessionClientSecret
+                        )
+                        .observe { [weak self] result in
+                            guard let self = self else { return }
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift:407]: [equivalentish web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/ba55e4ac/src/linkedAccounts/inner/hooks/v3/useSaveToNetworkAndLink.ts#L64-L71)

(note that this is missing code to set a success screen caption..will be added in different PR)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/AccountUpdateRequiredViewController.swift [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                primaryButtonConfiguration: PaneLayoutView.ButtonConfiguration(
+                    title: "Continue", // TODO: when Financial Connections starts supporting localization, change this to `String.Localized.continue`
+                    action: didSelectContinue
+                ),
+                secondaryButtonConfiguration: PaneLayoutView.ButtonConfiguration(
+                    title: "Cancel", // TODO: when Financial Connections starts supporting localization, change this to `String.Localized.cancel`
+                    action: didSelectCancel
+                )
+            ).footerView
+        )
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/AccountUpdateRequiredViewController.swift:65]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/containers/AccountUpdateRequiredDrawer.tsx#L77-L106)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 6]
```diff
@@ -217,47 +213,62 @@ final class LinkAccountPickerViewController: UIViewController {
...
         footerContainerView.addAndPinSubview(footerView)
 
-        bodyView.selectAccount(nil) // activate the logic to list all accounts
+        bodyView.selectAccounts([]) // activate the logic to list all accounts
     }
 
-    private func didSelectConectAccount() {
-        guard let selectedAccountTuple = dataSource.selectedAccountTuple else {
-            assertionFailure("user shouldn't be able to press the connect account button without an account")
+    private func didSelectConnectAccounts() {
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:219]: [equivalent web code for the function](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/containers/ReturningUser.tsx#L146-L176)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 7]
```diff
@@ -359,17 +351,163 @@ extension LinkAccountPickerViewController: LinkAccountPickerBodyViewDelegate {
...
+                    ],
+                    pane: .linkAccountPicker
+                )
+
+            dataSource.updateSelectedAccounts(
+                dataSource.selectedAccounts.filter(
+                    { $0.partnerAccount.id != selectedPartnerAccount.id }
+                )
+            )
+        }
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:382]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/containers/ReturningUser.tsx#L150-L160)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 8]
```diff
@@ -359,17 +351,163 @@ extension LinkAccountPickerViewController: LinkAccountPickerBodyViewDelegate {
...
+                )
+
+            if dataSource.manifest.singleAccount {
+                dataSource.updateSelectedAccounts([selectedAccountTuple])
+            }
+            // multi-select
+            else {
+                dataSource.updateSelectedAccounts(
+                    dataSource.selectedAccounts + [selectedAccountTuple]
+                )
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:403]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/containers/ReturningUser.tsx#L161-L173)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 9]
```diff
@@ -359,17 +351,163 @@ extension LinkAccountPickerViewController: LinkAccountPickerBodyViewDelegate {
...
+                        didSelectCancel: { [weak self] in
+                            delayDeselectingAccounts = false
+                            self?.lastAccountUpdateRequiredViewController?.dismiss(
+                                animated: true
+                            )
+                        },
+                        willDismissSheet: willDismissSheet
+                    )
+                    lastAccountUpdateRequiredViewController = accountUpdateRequiredViewController
+                    accountUpdateRequiredViewController.present(on: self)
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:507]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/lib/useUpdateRequiredDrawer.ts#L26-L92)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 10]
```diff
@@ -359,17 +351,163 @@ extension LinkAccountPickerViewController: LinkAccountPickerBodyViewDelegate {
...
+                        delayDeselectingAccounts = false
+                        self?.lastAccountUpdateRequiredViewController?.dismiss(
+                            animated: true
+                        )
+                    },
+                    willDismissSheet: willDismissSheet
+                )
+                lastAccountUpdateRequiredViewController = accountUpdateRequiredViewController
+                accountUpdateRequiredViewController.present(on: self)
+            }
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:509]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/lib/useUpdateRequiredDrawer.ts#L26-L92)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 11]
```diff
@@ -359,17 +351,163 @@ extension LinkAccountPickerViewController: LinkAccountPickerBodyViewDelegate {
...
+                    pane: .linkAccountPicker
+                )
+
+            dataSource.updateSelectedAccounts(
+                dataSource.selectedAccounts.filter(
+                    { $0.partnerAccount.id != selectedPartnerAccount.id }
+                )
             )
-        dataSource.updateSelectedAccount(selectedAccountTuple)
+        }
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:382]: [equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/containers/ReturningUser.tsx#L150-L160)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift [hunk 12]
```diff
@@ -387,11 +387,52 @@ final class AccountPickerViewController: UIViewController {
...
+                                        errorName: "SaveToLinkError",
+                                        pane: .accountPicker
+                                    )
+                            }
+
+                            self.delegate?.accountPickerViewController(
+                                self,
+                                didSelectAccounts: selectedAccounts,
+                                nextPane: .success
+                            )
```
Comment (bric-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift:423]: note sure if this is what you're saying in the above comment

> (note that this is missing code to set a success screen caption..will be added in different PR)

but as-is this will show success, even in the result == .failure case, right? intentional for now?

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 13]
```diff
@@ -217,47 +213,62 @@ final class LinkAccountPickerViewController: UIViewController {
...
-        bodyView.selectAccount(nil) // activate the logic to list all accounts
+        bodyView.selectAccounts([]) // activate the logic to list all accounts
     }
 
-    private func didSelectConectAccount() {
-        guard let selectedAccountTuple = dataSource.selectedAccountTuple else {
-            assertionFailure("user shouldn't be able to press the connect account button without an account")
+    private func didSelectConnectAccounts() {
+        let nextPane: FinancialConnectionsSessionManifest.NextPane? = {
+            if let mostRecentlySelectedAccount = dataSource.selectedAccounts.last {
```
Comment (bric-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:221]: just double checking: the `nextPaneOnSelection` should be the same for each account? so just grabbing the last works out?

**Review by bric-stripe — APPROVED** [review]

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift (hunk 12 above)
Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/Native/AccountPicker/AccountPickerViewController.swift:423]: sorry, the PR "referenced" is this one now:
- https://github.com/stripe/stripe-ios/pull/3459

the note probably was more confusing than helpful, but was there in case one analyzed the web code and saw missing code

> but as-is this will show success, even in the result == .failure case, right? intentional for now?

yes

this is indeed a bit confusing: one way to understand it is that "Financial Connections" could be thought of as having two layers:
1) There's the Financial Connections product that makes sure users can connect their bank accounts
2) The Link product is able to "super charge" this flow by saving bank accounts, and then later allowing consumers to connect their accounts quicker

This particular code does (2), but if it fails, its OK. User will still complete Financial Connections flow, but they won't have a quicker experience next time (...because account failed to save).

[web code of handling save to link error but going to success](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/3ac43f6e/src/linkedAccounts/inner/components/panes/v3/LinkLoginPane/lib/useHandleSaveToLinkError.ts#L48)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift (hunk 13 above)
Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:221]: yes - there's a bit of an assumption because backend controls what is returned (and they can change it) but that should be true

[this is the equivalent web code](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/f87b7a8a/src/linkedAccounts/inner/components/panes/v3/LinkAccountPickerPane/containers/ReturningUser.tsx#L107-L127) where they also pick the last account in the array

**Comment by kgaidis-stripe** [comment]
This PR has turned into a "feature branch" of other reviewed PR's 

PR 2 merged:
- https://github.com/stripe/stripe-ios/pull/3458

PR 3 merged here:
- https://github.com/stripe/stripe-ios/pull/3462

(apologies for the extra pings - hopefully will not be too many more)

**Review by bric-stripe — APPROVED** [review]
